### ISSUE TITLE
Competing archive and filename redirects for legacy release-notes links

### ISSUE THREAD

**Issue by abhinav-phi** [checkstyle/checkstyle#21619]
## Summary

Follow-up requested by @romani in merged PR #21573.

While validating that PR, @smita1078 reported that a valid legacy release-notes link reached its archive on production but reached the renamed release-notes index on the generated site report instead. The website JavaScript contains two independent redirect handlers that can select different destinations for the same legacy URL.

This report tracks that redirect interaction separately from the version-component guard fixed by #21571 / #21573. It concerns the documentation website, not a Java Checkstyle check; a Java input file, check configuration, and CLI audit are not applicable.

## Reported behavior and reproduction URLs

The validation report is in the [discussion on #21573](https://github.com/checkstyle/checkstyle/pull/21573). The site report was generated from PR head `74d6ce95d20728135ef53f6847a06b8b0f6e9a80`; the PR subsequently merged as `f444f28260361492b70eb34ceba59cfdbddb0061`.

To compare the reported behavior, open each legacy URL directly in a browser and inspect the final path and fragment:

| Input | Result reported during PR validation |
| --- | --- |
| [Production: `releasenotes.html#Release_6.0`](https://checkstyle.org/releasenotes.html#Release_6.0) | `release-notes-old-6-0-7-8.html#Release_6.0` — the expected archive |
| [Site report: `releasenotes.html#Release_6.0`](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/website/74d6ce9_20260915152417/releasenotes.html#Release_6.0) | `release-notes.html` rather than the archive |
| [Production: `releasenotes.html#Release_6`](https://checkstyle.org/releasenotes.html#Release_6) | `release-notes-old-6-0-7-8.html#Release_6` — the incomplete-version bug addressed by #21573 |
| [Site report: `releasenotes.html#Release_6`](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/website/74d6ce9_20260915152417/releasenotes.html#Release_6) | `release-notes.html` rather than an archive |

The valid `#Release_6.0` case is the focus of this issue. The incomplete `#Release_6` case is included to distinguish the merged guard fix from the remaining redirect interaction.

These are the reviewer's observations from validation, not a claim that every browser or every subsequent visit will reproduce the same destinations. Production may change after deployment, and the browser/version and cache conditions were not recorded in the original report. The rename handler preserves the fragment in its destination; a browser reproduction should record the complete final URL as well as the path.

## Relevant code

Both handlers below are present in the merged source, `src/site/resources/js/checkstyle.js`.

### 1. Archive redirect on `DOMContentLoaded`

[Source at the merge commit, lines 20–49](https://github.com/checkstyle/checkstyle/blob/f444f28260361492b70eb34ceba59cfdbddb0061/src/site/resources/js/checkstyle.js#L20-L49)

This handler processes `/releasenotes.html` when the fragment starts with `#Release_` and the current document does not already contain that anchor. After checking the version components, it selects an archive. For `#Release_6.0`, it calls:

```js
window.location.replace(`./release-notes-old-6-0-7-8.html${url.hash}`);
```

### 2. Filename redirect on `load`

[Source at the merge commit, lines 51–115](https://github.com/checkstyle/checkstyle/blob/f444f28260361492b70eb34ceba59cfdbddb0061/src/site/resources/js/checkstyle.js#L51-L115)

A separate handler processes pages whose title starts with `Redirecting`. Its map includes:

```js
"releasenotes.html": "release-notes.html",
```

It constructs another destination and requests navigation:

```js
const newUrl = `./${redirectMap[oldHtmlFile]}${urlObj.hash}`;
window.location.replace(newUrl);
```

Both fetched legacy pages had the title `Redirecting – checkstyle`. For a valid archived release, the first handler selects the archive, while the second can select the renamed index if the original document reaches its `load` handler before it is replaced. There is no explicit coordination between these redirect decisions.

## Evidence and limits of the diagnosis

During the PR investigation:

- Comparing the served production and site-report `checkstyle.js` files showed only the guard-line difference: `!versionParts.length >= 2` versus `versionParts.length < 2`.
- For `#Release_6.0`, `versionParts.length` is `2`. Neither version of that condition returns early, so the guard change does not change archive selection for this input.
- A Node DOM-stub harness executing the served scripts, with `DOMContentLoaded` followed by `load`, recorded an archive navigation request followed by a renamed-index navigation request for valid archived versions such as `#Release_7.8` and `#Release_8.20`, with both script versions.

The harness demonstrates the conflicting navigation requests when both handlers execute. It does **not** simulate browser navigation cancellation or establish which request commits in a real browser. Event/navigation timing is a plausible explanation for the reported discrepancy, but it still needs browser-level confirmation. The differing sizes of the fetched pages do not, by themselves, prove that page size caused the different outcomes.

## Expected behavior

A valid legacy archive link such as `releasenotes.html#Release_6.0` should reliably reach `release-notes-old-6-0-7-8.html#Release_6.0`. A generic filename redirect should not supersede an archive destination selected for that same URL.

Other existing navigation behavior should remain intact:

- Legacy links without an archive destination should retain the normal filename-rename fallback where appropriate.
- Incomplete versions such as `#Release_6` must not regain an archive redirect that #21573 removed.
- An existing release anchor in the current document must not be incorrectly routed to an archive.
- Unrelated documentation filename redirects should continue to work.

## Suggested direction and validation

Coordinate the archive and filename-rename decisions so that the intended destination is selected once, with explicit precedence for a matching archive. The implementation could centralize destination selection or otherwise prevent the generic redirect from superseding an archive request.

Simply skipping the rename for every `#Release_` fragment would be too broad: malformed, current, and otherwise unmatched versions still need defined fallback behavior. Moving handlers between events alone would also need verification that the competing navigation requests are actually eliminated.

Suggested regression coverage:

| Input category | Behavior to verify |
| --- | --- |
| Valid archived releases, including `6.0`, `7.8`, `8.20`, and `10.26` | Correct archive and unchanged fragment; no competing rename request |
| Incomplete version, e.g. `#Release_6` | No archive redirect; appropriate fallback remains available |
| No fragment | Normal `releasenotes.html` filename redirect |
| Anchor already present in the document | No erroneous archive selection |
| Version outside the archive ranges | Defined fallback, not an arbitrary archive |
| Other renamed documentation pages | Existing redirects remain unchanged |

Browser validation should record browser/version, final URL, and relevant cache/network conditions. Testing both production-like and generated-site pages would help establish that archive navigation no longer depends on the observed deployment difference.

## Related work and scope

- #21573: merged version-component guard fix and the discussion requesting this follow-up.
- #21571: original precedence bug; not being reopened here.
- #21398: broader proposal to centralize historical redirects and generate compatibility pages.

This issue is the narrower user-visible release-notes routing problem. It does not require the architectural migration proposed in #21398, although that work may provide the eventual solution. If addressed there, the archive-precedence and fallback cases above should be covered explicitly.

**Comment by jrdevadattan** [comment]
I can work on this docs-site redirect follow-up. I will focus on the two paths in `src/site/resources/js/checkstyle.js`: the archive redirect running on `DOMContentLoaded` and the filename redirect running on `load`. The main thing I want to preserve is archive precedence for real legacy fragments like `#Release_6.0`, `#Release_7.8`, `#Release_8.20`, and `#Release_10.26`, while still letting incomplete fragments such as `#Release_6` and plain `releasenotes.html` use the correct fallback. I will also browser-check the final URL and fragment so the fix does not only pass a unit-style check while still racing in the generated site.

**Comment by abhinav-phi** [comment]
@jrdevadattan - thank you for volunteering! Your scope is exactly right: archive precedence for real legacy fragments (#Release_6.0, #Release_7.8, #Release_8.20, #Release_10.26) in checkstyle.js, correct fallback for incomplete fragments like #Release_6 and plain releasenotes.html, plus a browser check of the final URL+fragment. I do not have permission to assign the issue, so a maintainer will need to do that - but please go ahead and open the PR whenever ready (referencing this issue), and I will review. Welcome!

**Comment by jiteshkhatri11** [comment]
I’m investigating the redirect interaction in `src/site/resources/js/checkstyle.js`, particularly how the `DOMContentLoaded` archive redirect and `load` filename redirect can issue competing navigations. I’ll first verify the existing site/JavaScript test coverage and the current fallback behavior before proposing a minimal fix.

**Comment by Adarsh-Me** [comment]
Hey @abhinav-phi , I investigated this issue and I think I found the main reason for the redirect problem.

In `src/site/resources/js/checkstyle.js`, there are two different redirect handlers working on `releasenotes.html`.

The first one runs on `DOMContentLoaded`. When we open something like `releasenotes.html#Release_6.0`, it correctly finds the version and redirects it to `release-notes-old-6-0-7-8.html#Release_6.0`.

But later, there is another handler running on `load`. It sees that the page is `releasenotes.html` and uses the `redirectMap`, where `releasenotes.html` is mapped to `release-notes.html`. So in some cases both handlers can call `window.location.replace()`. Because of this, the final URL can depend on which navigation happens first/gets applied.

I think the safest fix is to give the archive redirect priority instead of trying to fix this by changing `DOMContentLoaded` to `load` or changing the event order.

Basically, in the generic filename redirect, we should skip the `releasenotes.html -> release-notes.html` redirect when the URL has a valid release version which should go to an archive. This way the archive handler remains responsible for those old release links.

We should not skip `releasenotes.html` for every `#Release_` hash though. For example, `#Release_6` is an incomplete version and should not be sent to an archive. Also, a normal `releasenotes.html` URL without a hash should still redirect to `release-notes.html`.

I would test at least these cases after the change:

* `#Release_6.0` → `release-notes-old-6-0-7-8.html`
* `#Release_7.8` → `release-notes-old-6-0-7-8.html`
* `#Release_8.20` → `release-notes-old-8-0-8-34.html`
* `#Release_10.26` → `release-notes-old-8-35-10-26.html`
* `#Release_6` → no archive redirect
* no hash → `release-notes.html`
* existing release anchor → no unwanted archive redirect
* other old documentation URLs → existing redirects should still work

I think this is better than only changing the event timing because it actually removes the two redirects fighting with each other. I would also do a browser test on the generated site report to make sure the final URL and hash are correct.

**Comment by abhinav-phi** [comment]
Hello everyone, and thank you all for picking interest in this issue. Since three parallel directions have now emerged, I want to set one clear direction so that efforts don't split and reviewers don't end up comparing competing fixes.

**Decision:** we will continue with @Adarsh-Me's approach — and @Adarsh-Me, please go ahead on your approach and open the PR as soon as possible.

**Why this one (same bar for everyone):** the deciding factors are concreteness and verifiability — an exact root cause, a precise fix mechanism, the exact version→archive mapping, and a test matrix plus a real browser check on the generated site. @Adarsh-Me's proposal is the only one that meets all four today.

**A correction from my side:** in my earlier comment I endorsed @jrdevadattan's scope as exactly right. That was premature of me — I was not correct. A scope outline, however well directed, is not sufficient to proceed on without the concrete mechanism behind it, so I am withdrawing that endorsement.

**What this means for everyone:** @jrdevadattan and @jiteshkhatri11 — thank you for stepping up, but please don't continue separate implementations of this issue. If you'd like to help, the most useful contribution now is reviewing and testing @Adarsh-Me's PR once it is up (especially the browser-level final-URL check). Converging on one implementation is the fastest way to get this fixed and merged.

**Comment by Adarsh-Me** [comment]
> Hello everyone, and thank you all for picking interest in this issue. Since three parallel directions have now emerged, I want to set one clear direction so that efforts don't split and reviewers don't end up comparing competing fixes.
> 
> **Decision:** we will continue with [@Adarsh-Me](https://github.com/Adarsh-Me)'s approach — and [@Adarsh-Me](https://github.com/Adarsh-Me), please go ahead on your approach and open the PR as soon as possible.
> 
> **Why this one (same bar for everyone):** the deciding factors are concreteness and verifiability — an exact root cause, a precise fix mechanism, the exact version→archive mapping, and a test matrix plus a real browser check on the generated site. [@Adarsh-Me](https://github.com/Adarsh-Me)'s proposal is the only one that meets all four today.
> 
> **A correction from my side:** in my earlier comment I endorsed [@jrdevadattan](https://github.com/jrdevadattan)'s scope as exactly right. That was premature of me — I was not correct. A scope outline, however well directed, is not sufficient to proceed on without the concrete mechanism behind it, so I am withdrawing that endorsement.
> 
> **What this means for everyone:** [@jrdevadattan](https://github.com/jrdevadattan) and [@jiteshkhatri11](https://github.com/jiteshkhatri11) — thank you for stepping up, but please don't continue separate implementations of this issue. If you'd like to help, the most useful contribution now is reviewing and testing [@Adarsh-Me](https://github.com/Adarsh-Me)'s PR once it is up (especially the browser-level final-URL check). Converging on one implementation is the fastest way to get this fixed and merged.

Hey @abhinav-phi please review my pr i have appilied the changes as we discussed in this thread.
